src/System/Terminal/Widgets/SearchSelect.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE StrictData #-}
+
+module System.Terminal.Widgets.SearchSelect where
+
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Text qualified as T
+import Data.Text.Rope.Zipper qualified as RopeZipper
+import Internal.Prelude
+import System.Terminal.Render qualified as Render
+import System.Terminal.Widgets.Common
+import System.Terminal.Widgets.Select
+import System.Terminal.Widgets.TextInput
+
+-- | SearchSelect is a combination of 'TextInput' and 'Select'; it allows the
+-- selections to be searched and sorted via the user's text input, similar to
+-- [vertico for Emacs](https://github.com/minad/vertico).
+--
+-- Note: FIXME: A 'Map' is used internally. Therefore, the order of the search
+-- result is determined by the output of 'Data.Map.toList'. For as long as 'Select'
+-- continues to use a list, it may be preferable to use the converted output
+-- to initialize 'selectWidget'. This prevents the jarring transition upon
+-- first user input.
+--
+-- Example:
+-- >>> textInput = TextInput "M-x " False False mempty id
+-- >>> opts = Map.toList $ Map.fromList [("foo", False), ("bar", True), ("baz", False)]
+-- >>> select = Select "Options" opts False 0
+-- >>> searchSelect = SearchSelect textInput select (Map.fromList opts) True 10
+-- >>> withTerminal . runTerminalT $ runWidget searchSelect
+-- @
+-- M-x <search>
+-- Options
+--  (*) [bar]
+--  ( ) baz
+--  ( ) foo
+-- @
+data SearchSelect = SearchSelect
+    { searchWidget :: TextInput
+    -- ^ The field the user uses to search the options.
+    , selectWidget :: Select
+    -- ^ The options presented to a user. Searched through via the searchField.
+    -- This widget should /not/ be used to query all available options;
+    -- it only shows the filtered output.
+    , globalOptions :: Map Text Bool
+    -- ^ Used internally to keep track of all options and their statuses, even
+    -- when not shown to the user.
+    , showOnNoInput :: Bool
+    -- ^ Whether to show the options when the user has yet to type any input.
+    , showCount :: Int
+    -- ^ The maximum number of items to show at any given time.
+    }
+    deriving stock (Generic)
+
+filterSS :: Text -> SearchSelect -> SearchSelect
+filterSS t ss@(SearchSelect{selectWidget, globalOptions, showCount}) =
```

Comment:
When taking multiple things from `ss` like this, while also needing `ss` itself, avoid pattern matching and just use `ss.` below.